logging/net_appender_test.go

```diff
@@ -130,3 +130,51 @@ func TestNetLoggerBatchWrites(t *testing.T) {
 		test.That(t, server.service.logs[i].Message, test.ShouldEqual, "Some-info")
 	}
 }
+
+func TestNetLoggerBatchFailureAndRetry(t *testing.T) {
+	server := makeServerForRobotLogger(t)
+	defer server.stop()
+
+	netAppender, err := NewNetAppender(server.cloudConfig)
+	test.That(t, err, test.ShouldBeNil)
+	logger := NewDebugLogger("test logger")
+	// The stdout appender is not necessary for test correctness. But it does provide information in
+	// the output w.r.t the injected grpc errors.
+	logger.AddAppender(netAppender)
+
+	// This test will first log 10 "Some-info" logs. Followed by a single "New info" log.
+	numLogs := 11
+
+	// Inject a failure into the server handling `Log` requests.
+	server.service.logsMu.Lock()
+	server.service.logFailForSizeCount = numLogs
+	server.service.logsMu.Unlock()
+
+	for i := 0; i < numLogs-1; i++ {
+		logger.Info("Some-info")
+	}
+
+	// This test requires at least three syncs for the logs to be guaranteed received by the
+	// server. Once the log queue is full of size ten batches, the first sync will decrement
+	// `logFailForSizeCount` to 1 and return an error. The second will decrement it to a negative
+	// value and return an error. The third will succeed.
```

Comment:
minor - should we add this third `Sync()` to the the test and assert that the logs are flushed? or is that duplicative of other tests?

Reply:
Ah good q; you'll see in a comment below this that we rely on the `Close` call waiting for a bit and a `syncOnce()` happening asynchronously, so we don't ned to call `Sync()` _three_ times.

Reply:
ah I overlooked this, the flow makes perfect sense now.